File: ArraysHashing/2Sum/Optimal/2sumo.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

vector<pair<int,int>> twoSum(vector<int>& array,int result)
{
    unordered_map<int,int> map;
    vector<pair<int,int>> pairs;
    int complement = 0;

    for(int i = 0; i < array.size(); i++)
    {
        complement = result - array[i];

    // map[complement] > 0 ensures that in case of duplicates pairing is handled properly

        if(map.find(complement) != map.end() && map[complement] > 0)
        {
            pairs.push_back({complement,array[i]});
            map[complement] --;
        }
        else
        {
        map[array[i]]++;
        }
    }

    return pairs;
}

int main()
{
    vector<int> array = {2,7,11,15,2,7,3,8};
    int target = 9;

    vector<pair<int,int>> result = twoSum(array,target);

    if(!result.empty())
    {
        cout << "Pairs found : " << endl;
        for(const auto& p:result)
        {
            cout << "(" << p.first << ", " << p.second << ")" << endl;
        }
    }
    else{
        cout << "No pairs found." << endl;
    }

    return 0;
}

File: ArraysHashing/2Sum/Optimal/2sumo.py
from typing import List,Tuple,Dict

def two_sum_all_pairs(nums : List[int],target : int) -> List[Tuple[int,int]] :
    map : Dict[int,int] = {}
    result : List[Tuple[int,int]] = []
    
    for num in nums:
        complement : int = target - num
        
        if complement in map and map[complement] > 0:
            result.append((complement,num))
            map[complement] -= 1
        else:
            map[num] = map.get(num,0) + 1
    
    return result

nums: List[int] = [2,7,11,15,2,7,3,8]
target :int = 9
result : List[Tuple[int,int]] = two_sum_all_pairs(nums,target)

if result:
    print("Found pairs : ")
    for pair in result:
        print(f"({pair[0]},{pair[1]})")

else:
    print("No pairs found")